### PR TITLE
Implement folo listeners

### PR DESCRIPTION
**Pull request by hjmodi**

### CODE REVIEW

**Review by ligangty — COMMENTED** [review]
@hjmodi Some comments in code.
And I've confirmed with @sswguo that current indy-monolith does not implement the FileAccessEvent publishing, so maybe you need to take care of this and add some code in it for this part of functions.

File: src/main/java/org/commonjava/indy/service/tracking/change/FoloTrackingAdjustListener.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            logger.trace( "No sealed record found, trackingKey: {}", trackingKey );
+            return;
+        }
+
+        adjustTrackedContent( trackedContent, source, target );
+
+        recordManager.replaceTrackingRecord( trackedContent );
+    }
+
+    private void adjustTrackedContent( TrackedContent trackedContent, StoreKey source, StoreKey target )
```
[the remark this answers is not in the document]
Comment (hjmodi) [reply src/main/java/org/commonjava/indy/service/tracking/change/FoloTrackingAdjustListener.java]: removed it

File: src/main/java/org/commonjava/indy/service/tracking/change/FoloTrackingListener.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            handleFileAccessEvent( event );
+        }
+        else if ( event.getEventType().equals( FileEventType.STORAGE ) )
+        {
+            handleFileStorageEvent( event );
+        }
+
+    }
+
+    private TrackedContentEntry createEntry( final TrackingKey trackingKey, final StoreKey affectedStore,
```
[the remark this answers is not in the document]
Comment (hjmodi) [reply src/main/java/org/commonjava/indy/service/tracking/change/FoloTrackingListener.java]: removed it

File: src/main/java/org/commonjava/indy/service/tracking/change/FoloTrackingListener.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    {
+        logger.trace( "FILE STORAGE: {}", event );
+
+        EventMetadata metadata = event.getEventMetadata();
+
+        logger.warn( ">>> Metadata: " + metadata );
+
+        String trackingId = event.getSessionId();
+
+        final TrackingKey trackingKey = !Objects.equals( trackingId, "" ) ? new TrackingKey( trackingId ) : null;
```
[the remark this answers is not in the document]
Comment (hjmodi) [reply src/main/java/org/commonjava/indy/service/tracking/change/FoloTrackingListener.java]: good idea. Using it now.

**Comment by hjmodi** [comment]
> @hjmodi Some comments in code. And I've confirmed with @sswguo that current indy-monolith does not implement the FileAccessEvent publishing, so maybe you need to take care of this and add some code in it for this part of functions.

yes, would you also know how the promote events are sent to kafka?

**Review by ligangty — APPROVED** [review]

**Comment by ligangty** [comment]
> > @hjmodi Some comments in code. And I've confirmed with @sswguo that current indy-monolith does not implement the FileAccessEvent publishing, so maybe you need to take care of this and add some code in it for this part of functions.
> 
> yes, would you also know how the promote events are sent to kafka?

The promote events are sent through indy-promote-service after Indy 3.1.0 deployment. You can check https://github.com/Commonjava/indy-promote-service/blob/main/src/main/java/org/commonjava/service/promote/client/kafka/KafkaEventDispatcher.java and its usage in other code for details.

**Comment by hjmodi** [comment]
@ligangty here is the PR for indy https://github.com/Commonjava/indy/pull/2261

**Pull request merged**